BigInt.c

```diff
@@ -1,69 +1,108 @@
 #include <assert.h>
 #include <math.h>
+#include <errno.h>
+#include <limits.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "BigInt.h"
+#include "safe_math_impl.h"
 
 #define MAX(x, y) ((x) > (y) ? (x) : (y))
 
+#if UNIT_MAX >> 32 == 0
+#	define check_add_int_int check_add_int32_int32
+#	define check_mul_int_int check_mul_int32_int32
+#else
+#	if UNIT_MAX >> 64 == 0
+#		define check_add_int_int check_add_int64_int64
+#		define check_mul_int_int check_mul_int64_int64
+#	else
+#		error unsupported integer size
+#	endif
+#endif
+
 BigInt* BigInt_construct(int value) {
 
+    errno = 0;
     BigInt* new_big_int = malloc(sizeof(BigInt));
-    
+    if(!new_big_int) {
+        if(!errno) {
+            errno = ENOMEM;
+        }
+        return NULL;
+    }
+    unsigned int value2;
     if(value < 0) {
         new_big_int->is_negative = 1;
-        value *= -1;
+        value2 = -value;
     } else {
         new_big_int->is_negative = 0;
+        value2 = value;
     }
 
-    new_big_int->num_digits = floor(log10(value)) + 1;
+    new_big_int->num_digits = floor(log10(value2)) + 1;
 
     // Special case for 0
     if(new_big_int->num_digits == 0) {
         new_big_int->num_digits = 1;
     }
 
     new_big_int->num_allocated_digits = new_big_int->num_digits;
+    errno = 0;
     new_big_int->digits = malloc(new_big_int->num_allocated_digits * sizeof(unsigned char));
+    if(!new_big_int->digits) {
+        free(new_big_int);
+        if(!errno) {
+            errno = ENOMEM;
+        }
+        return NULL;
+    }
 
-    int i;
-    for(i = 0; i < new_big_int->num_digits; i++) {
-        new_big_int->digits[i] = value % 10;
-        value /= 10;
+    unsigned int count = new_big_int->num_digits;
+    char* digits = new_big_int->digits;
```

Comment:
```suggestion
    unsigned char* digits = new_big_int->digits;
```
gcc emits a `-Wpointer-sign` warning otherwise.